ttnn/cpp/ttnn/tensor/tensor_utils.cpp

```diff
@@ -213,6 +213,234 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+}
+
+template <typename T>
+Tensor to_bias_tile_layout_block_sharded(
+    const Tensor& conv_bias_tensor, uint32_t num_channel_shards, DataType output_dtype) {
+    auto b_shape = conv_bias_tensor.get_legacy_shape();
+    TT_ASSERT(b_shape[0] == 1 && b_shape[1] == 1 && b_shape[2] == 1);
+    auto compute = [&b_shape, &num_channel_shards, &output_dtype](const auto& input_buffer) {
+        auto bias_matrix_cols = b_shape[3];
+        /*TT_ASSERT(bias_matrix_cols % num_channel_shards == 0);*/
+        auto conv_output_shard_width = bias_matrix_cols / num_channel_shards;
+        auto conv_output_shard_width_padded =
+            (uint32_t)std::ceil((double)conv_output_shard_width / (double)constants::TILE_WIDTH) *
+            constants::TILE_WIDTH;
+        if (conv_output_shard_width < conv_output_shard_width_padded) {
+            // width padding for conv output shard padding
+            bias_matrix_cols = conv_output_shard_width_padded * num_channel_shards;
+        }
+
+        auto bias_matrix_rows = 32;
+        tt::tt_metal::LegacyShape output_shape = {1, 1, bias_matrix_rows, bias_matrix_cols};
+        auto output_buffer = owned_buffer::create<T>(compute_volume(output_shape));
+        for (auto oc = 0; oc < num_channel_shards; oc++) {
+            for (auto k_s = 0; k_s < conv_output_shard_width; k_s++) {
+                auto matrix_idx = oc * conv_output_shard_width_padded + k_s;
+                auto idx = oc * conv_output_shard_width + k_s;
+                output_buffer[matrix_idx] = input_buffer[idx];
+            }
+        }
+        if constexpr (std::is_same<T, float>::value) {
+            if (output_dtype == DataType::BFLOAT8_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp8_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+            if (output_dtype == DataType::BFLOAT4_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp4_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
```

Comment:
This looks like the same code copied/pasted again and again in this file. Can you please refactor this?

Reply:
Done.